### PR TITLE
Feat/button loading

### PR DESCRIPTION
**Pull request by dominguesgm**
Added `loading`, `success` and `error` to the button component.

Closes https://github.com/ipfs-shipyard/pm-idm/issues/73 .

Also changes the approach through which we import svg files. These are now imported inline so that one can change the svg's presentation attributes through css.

### CODE REVIEW

File: src/components/button/Button.js [hunk 1]
```diff
@@ -1,30 +1,160 @@
...
+
+    componentWillUnmount() {
+        this.clearFeedbackOutcomeTimers();
+    }
+
     render() {
         const { variant, disabled, children, fullWidth, ...rest } = this.props;
+        const { feedback, feedbackOutcome } = this.state;
+        const loading = feedback === 'loading';
+        const hasFeedback = feedback === 'loading' || feedback === 'success' || feedback === 'error';
```
Comment (paulobmarcos) [inline src/components/button/Button.js]: Correct me if I misunderstood this, but wouldn't it be the same for state feedback to start with `none` and simply check:

`const hasFeedback = feedback !== 'none';`

File: src/components/button/Button.js [hunk 2]
```diff
@@ -1,30 +1,160 @@
...
+                    <CheckmarkIcon className={ styles.checkmark } onTransitionEnd={ this.handleSuccessIconTransitionEnd } />
+                </span>
+                <span className={ errorBlockClassName } onTransitionEnd={ this.handleSpanTransitionEnd }>
+                    <CrossIcon className={ styles.cross } onTransitionEnd={ this.handleErrorIconTransitionEnd } />
+                </span>
+            </div>
         );
     }
+
+    maybeHandleFeedbackChange(prevFeedback) {
```
Comment (paulobmarcos) [inline src/components/button/Button.js]: `maybe` seems kind of strange 😕should we think of a better name?

I think we should call it `handleFeedbackChange`. The verification would be done in `componentDidUpdate` only since the `componentDidMount` actually wants to "handle" the change, right?

File: src/components/button/Button.js [hunk 3]
```diff
@@ -1,30 +1,160 @@
...
+        );
+
+        const wrapperClassName = classNames(
+            styles.buttonWrapper,
+            styles[feedbackOutcome],
+            styles[variant],
+        );
+
+        const propsFeedback = this.props.feedback;
+        const hasPropsFeedback = propsFeedback === 'loading' || propsFeedback === 'success' || propsFeedback === 'error';
```
Comment (paulobmarcos) [inline src/components/button/Button.js]: Same here? Since the the `feedback` prop can only be one of `['none', 'loading', 'success', 'error']`?

Comment (paulobmarcos) [reply src/components/button/Button.js]: Also this could remove the need of `propsFeedback`, like so: `const hasPropsFeedback = this.props.feedback !== 'none';`

File: .storybook/webpack.config.js [hunk 4]
```diff
@@ -1,10 +1,12 @@
 const path = require('path');
 const SvgStorePlugin = require('external-svg-sprite-loader');
 
-module.exports = (config) => {
+module.exports = ( config ) => {
```
Comment (PedroMiguelSS) [inline .storybook/webpack.config.js]: What's the reason for this change?

File: package.json [hunk 5]
```diff
@@ -23,28 +23,40 @@
   "scripts": {
     "start": "start-storybook -p 3010 -c .storybook",
     "lint": "eslint --ignore-path .gitignore . && stylelint --ignore-path .gitignore \"**/*.css\"",
+    "build": "npm run build:clean && npm run build:js && npm run build:css && npm run build:copy",
+    "build:clean": "rimraf dist styles",
```
Comment (PedroMiguelSS) [inline package.json:31]: I think [rimraf](https://github.com/isaacs/rimraf) is not installed.

File: src/components/button/Button.js [hunk 6]
```diff
@@ -1,30 +1,160 @@
...
+    handleIconTransitionEnd = (event, isSuccess) => {
+        event.stopPropagation();
+        if (this.enterAnimation) {
+            this.props.onAnimationEnd && this.props.onAnimationEnd(isSuccess);
+        }
+    };
+
+    handleSuccessIconTransitionEnd = (e) => this.handleIconTransitionEnd(e, true);
+
+    handleErrorIconTransitionEnd = (e) => this.handleIconTransitionEnd(e, true);
```
Comment (PedroMiguelSS) [inline src/components/button/Button.js]: As far as I understood what you're doing here, in this case, `this.handleIconTransitionEnd` must be called with its second argument as `false` instead of `true`. Am I right?

File: src/components/button/Button.js [hunk 7]
```diff
@@ -1,30 +1,160 @@
...
     variant: PropTypes.oneOf(['primary', 'secondary', 'tertiary', 'negative']).isRequired,
     disabled: PropTypes.bool,
     fullWidth: PropTypes.bool,
+    feedback: PropTypes.oneOf(['none', 'loading', 'success', 'error']),
+    onAnimationEnd: PropTypes.func,
     children: PropTypes.node.isRequired,
 };
 
+Button.defaultProps = {
+    feedback: PropTypes.oneOf(['none', 'loading', 'success', 'error']),
```
Comment (PedroMiguelSS) [inline src/components/button/Button.js]: You must choose one of the possible values to set as a default value for this prop.

File: src/components/button/ProgressBar.js [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+        // Skip if `running` hasn't changed
+        if (running === prevRunning) {
+            return;
+        }
+
+        if (running) {
+            this.begin();
+        } else {
+            this.end();
+        }
```
Comment (PedroMiguelSS) [inline src/components/button/ProgressBar.js]: I would use the ternary operator here, like this: 
```js
running ? this.begin() : this.end()
```

But this is obviously a personal preference.

File: src/components/button/ProgressBar.js [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+import styles from './ProgressBar.css';
+
+const INCREMENT_INTERVAL = 300;
+const MAX_PROGRESS = 0.95;
+
+class ProgressBar extends Component {
+    progressBarRef = createRef();
+
+    componentDidMount() {
+        this.maybeHandleRunningChange(false);
```
Comment (paulobmarcos) [inline src/components/button/ProgressBar.js]: If I understood correctly, the following would be better to understand: `this.props.running && this.begin();`

File: src/components/button/ProgressBar.js [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+    }
+
+    componentDidUpdate(prevProps) {
+        this.maybeHandleRunningChange(prevProps.running);
```
Comment (paulobmarcos) [inline src/components/button/ProgressBar.js]: Same as the Button component, should we change it to `handleRunningChange` ?

File: src/components/button/README.md [hunk 11]
```diff
@@ -17,6 +17,8 @@ import { Button } from '@nomios/web-uikit';
 | variant | string | *required* | Sets the button variant. Can be one of: `primary`, `secondary` `tertiary` or `negative` |
 | disabled | boolean | false | Sets the disable state |
 | fullWidth | boolean | false | Sets button width to 100% |
+| feedback | string | | Sets the current feedback of the button. Can be one of: `none`, `loading`, `success`, `error` |
+| onAnimationEnd | function | | Function called after the `success` or `error` entrance animation finishes. Takes a boolean `isSuccess` as an argument: `true` if the animation occurs after a `success` feedback and `false` otherwise |
```
Comment (PedroMiguelSS) [inline src/components/button/README.md:21]: This `onAnimationEnd` description proves that I was right on my comment above! 😁

File: src/components/icon/Icon.js [hunk 12]
```diff
@@ -1,41 +1,53 @@
...
+        super(props);
+
+        if (typeof props.svg === 'string') {
+            this.state = {
+                contents: props.svg,
+            };
+        } else {
+            this.state = {
+                contents: null,
+            };
```
Comment (paulobmarcos) [inline src/components/icon/Icon.js]: We can avoid the if statement here with:

```js
this.state = {
   contents: typeof props.svg === 'string' ? props.svg : null,
}
```

**Review by paulobmarcos — CHANGES_REQUESTED** [review]

File: src/components/icon/Icon.js (hunk 12 above)
Comment (PedroMiguelSS) [inline src/components/icon/Icon.js]: I would change this to:
```js
this.state = {
    contents: typeof props.svg === 'string' ? props.svg : null
}
```

or 

```js
const contents = typeof props.svg === 'string' ? props.svg : null;

this.state = { contents };
```

**Review by PedroMiguelSS — CHANGES_REQUESTED** [review]

File: src/components/button/ProgressBar.js (hunk 8 above)
Comment (PedroMiguelSS) [reply src/components/button/ProgressBar.js]: Oh, wait! I think the linter will complain about this approach. But you can test it anyway.

File: src/components/button/Button.js (hunk 1 above)
Comment (dominguesgm) [reply src/components/button/Button.js]: You're right, I just needed to set 'none' as the default value for the feedback prop

File: src/components/button/Button.js (hunk 2 above)
Comment (dominguesgm) [reply src/components/button/Button.js]: Personally I think so too. This is part of the logic that was borrowed from discussify and is named as such because there isn't necessarily a feedback change when this function is called. 

I've moved this verification to the functions that called `maybeHandleFeedbackChange` and renamed it as suggested

File: src/components/button/Button.js (hunk 3 above)
Comment (dominguesgm) [reply src/components/button/Button.js]: 👍

File: src/components/button/ProgressBar.js (hunk 9 above)
Comment (dominguesgm) [reply src/components/button/ProgressBar.js]: I'd prefer leaving as is, as the behavior is used multiple times and is already abstracted in a function. You'd have to consider the execution of `this.end()` as well, in case `running`starts off as false, and to have that if/else format, we might as well keep the function we already have.

File: src/components/button/ProgressBar.js (hunk 10 above)
Comment (dominguesgm) [reply src/components/button/ProgressBar.js]: 👍

File: src/components/icon/Icon.js (hunk 12 above)
Comment (dominguesgm) [reply src/components/icon/Icon.js]: 👍

File: .storybook/webpack.config.js (hunk 4 above)
Comment (dominguesgm) [reply .storybook/webpack.config.js]: None. They're remnants of trying to upgrade storybook's version to v5. Changing it back to what it was.

File: package.json (hunk 5 above)
Comment (dominguesgm) [reply package.json:31]: you're right, for some reason I didn't notice any warning about it during builds.

File: src/components/button/Button.js (hunk 6 above)
Comment (dominguesgm) [reply src/components/button/Button.js]: Yes, nice catch 👍

File: src/components/button/Button.js (hunk 7 above)
Comment (dominguesgm) [reply src/components/button/Button.js]: Fixed it while making another change requested by paulo, 👍

File: src/components/button/ProgressBar.js (hunk 8 above)
Comment (dominguesgm) [reply src/components/button/ProgressBar.js]: Yup, it doesn't accept an expression in that context.

File: src/components/button/README.md (hunk 11 above)
Comment (dominguesgm) [reply src/components/button/README.md:21]: Hahaha yes! It was probably an issue caused by mindlessly copying the line of code and forgeting to change that argument 😂

File: src/components/icon/Icon.js (hunk 12 above)
Comment (dominguesgm) [reply src/components/icon/Icon.js]: @paulobmarcos suggested the same thing. I changed it to your first alternative 👍

File: src/components/button/ProgressBar.js (hunk 9 above)
Comment (paulobmarcos) [reply src/components/button/ProgressBar.js]: Seems fine to me 😁

**Review by paulobmarcos — APPROVED** [review]

**Review by PedroMiguelSS — APPROVED** [review]
